mozfldp/simulation_runner.py

```diff
@@ -0,0 +1,143 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from mozfldp.client import Client
+
+# TODO: how should this be invoked?
+from mozfldp.server import start_server
+
+import numpy as np
+
+
+def _format_data_for_model(dataset, label_col, user_id_col):
+    """Split a DataFrame into feature & label arrays."""
+    if not dataset:
+        return (None, None)
+    if user_id_col in dataset.columns:
+        dataset = dataset.drop(columns=user_id_col)
+    return (np.asarray(dataset.drop(columns=label_col)), np.asarray(dataset[label_col]))
+
+
+class BaseSimulationRunner:
+    """Manager for the server/client simulation infrastructure.
+
+    On initialization, the data is allocated across clients.
+    Individual simulation rounds can then be performed using
+    `run_simulation_round()`.
+
+    model: a `SGDModel` instance to be used for the simulation
+    training_data: a DataFrame containing the full training data. It should have
+        columns indicating a user ID and training labels. All other columns are
+        considered to be features
+    coef_init: the initial choice for the model coefficients
+    intercept_init: the initial choice for the model intercept
+    test_data: a DataFrame in the same format as `training_data`
+    label_col: name of the column containing labels
+    user_id_col: name of the column containing user IDs
+    """
+
+    def __init__(
+        self,
+        model,
+        training_data,
+        coef_init,
+        intercept_init,
+        test_data=None,
+        label_col="labels",
+        user_id_col="user_id",
+    ):
+        self._model = model
+        # Make the model aware of the full set of labels (necessary for partial
+        # fit updating).
+        self._model.set_training_classes(training_data[label_col])
+        # Maintain the history of model weights for each round.
+        self._coefs = [coef_init]
+        self._intercepts = [intercept_init]
+        self._num_rounds_completed = 0
+
+        # Initialize the clients with their respective datasets.
+        self._clients = []
+        by_user_data = training_data.groupby("user_id")
+        for user_id, user_data in by_user_data:
+            feats, labs = _format_data_for_model(user_data, label_col, user_id_col)
+            self._clients.append(Client(user_id, feats, labs, self._model.get_clone()))
```

Comment:
Shouldn't this `get_clone` call require `true` to use a pretrained model?

Reply:
I think you're right. I was originally thinking to initialize clients with an untrained model, but this should probably just pull in whatever state the supplied model is in. Ie. to start a fresh simulation, the caller should take care to pass in an untrained model.